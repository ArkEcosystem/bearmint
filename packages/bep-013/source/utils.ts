export interface NonNullableMap<T> {
	all(): Record<string, T>
	entries(): Array<[string, T]>
	get(key: string): NonNullable<T>
	has(key: string): boolean
	keys(): string[]
	missing(key: string): boolean
	set(key: string, value: T): void
	values(): T[]
}
