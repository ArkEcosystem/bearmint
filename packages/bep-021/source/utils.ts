import { assert } from '@bearmint/bep-009'
import type { Milestone } from '@bearmint/bep-013'

export function getPublicKeyType(milestone: Milestone) {
	const pubKeyTypes = milestone.parameters.consensus.validator?.pubKeyTypes

	assert.defined<string[]>(pubKeyTypes)
	assert.defined<string[]>(pubKeyTypes[0])

	return pubKeyTypes[0]
}
