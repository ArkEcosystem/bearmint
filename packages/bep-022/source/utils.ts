import type { AccountWithValidator, GenesisValidator } from '@bearmint/bep-013'

export function isGenesis(validators: GenesisValidator[], account: AccountWithValidator) {
	return validators.some(({ address }) => address === account.validator.address)
}
