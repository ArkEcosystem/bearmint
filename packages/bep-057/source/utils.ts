import {
	deriveEd25519Address,
	deriveSecp256k1Address,
	deriveSr25519Address,
} from '@bearmint/bep-089'

export function deriveAddress(publicKey: Uint8Array, publicKeyType: string) {
	if (publicKeyType === 'ed25519') {
		return deriveEd25519Address(publicKey)
	}

	if (publicKeyType === 'sr25519') {
		return deriveSr25519Address(publicKey)
	}

	if (publicKeyType === 'secp256k1') {
		return deriveSecp256k1Address(publicKey)
	}

	throw new Error(`Public Key type (${publicKeyType}) is not supported.`)
}
