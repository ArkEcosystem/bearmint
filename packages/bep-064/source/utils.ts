import { assert, blake3, bytesToHex, bytesToJson } from '@bearmint/bep-009'
import type { KVStore, TxFactory, TxMsgFactoryRegistry } from '@bearmint/bep-013'

import type { CustomTxMsgFactory } from './tx-msg.js'
import type { PendingLock } from './types.js'
import { HANDLER } from './types.js'

export function getPrimaryKey(hash: string, index: string) {
	return bytesToHex(blake3(`${HANDLER}/h/${hash}/i/${index}`))
}

export async function getPendingLock(
	txStore: KVStore,
	txFactory: TxFactory,
	txMsgServiceRegistry: TxMsgFactoryRegistry,
	bep64: KVStore,
	hash: string,
): Promise<PendingLock> {
	// 1. get the hash of the htlc lock tx based on the randomly generated lock hash
	const pendingLock = bytesToJson<{ hash: string; index: number }>(await bep64.get(hash))

	// 2. get the htlc lock tx
	const bytes = await txStore.get(pendingLock.hash)

	// 3. decode the htlc lock tx
	const tx = await txFactory.fromBytes(bytes)

	// 4. canonicalize the htlc lock tx
	const { ops } = await txMsgServiceRegistry.getFromTx<CustomTxMsgFactory>(tx).fromTx(tx)

	// 5. grab the lock at the exact index
	const lock = ops[Number(pendingLock.index)]

	assert.defined(lock)

	return {
		...lock,
		sender: tx.data.sender.toString(),
	}
}
