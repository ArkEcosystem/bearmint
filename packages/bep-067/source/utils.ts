import { blake3, bytesToHex } from '@bearmint/bep-009'

import { HANDLER } from './types.js'

export function getPrimaryKey(type: string, value: string) {
	return bytesToHex(blake3(`${HANDLER}/t/${type}/v/${value}`))
}
