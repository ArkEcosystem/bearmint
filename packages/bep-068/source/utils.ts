import { blake3, bytesToHex } from '@bearmint/bep-009'

import { HANDLER } from './types.js'

export function getPrimaryKey(hash: Uint8Array) {
	return bytesToHex(blake3(`${HANDLER}/d/${bytesToHex(hash)}`))
}
