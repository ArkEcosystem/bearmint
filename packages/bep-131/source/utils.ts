import { blake3, bytesToHex } from '@bearmint/bep-009'
import type { KVStore } from '@bearmint/bep-013'
import { PROTO_OPTS } from '@bearmint/bep-013'
import type { BEP126Collection, Plain } from '@bearmint/bep-018'
import { BEP131Token } from '@bearmint/bep-018'

import { HANDLER } from './types.js'

export function makeTokenId(collection: BEP126Collection) {
	return Object.keys(collection.tokens).length + 1
}

export function getPrimaryKey(collectionId: string, tokenId: number) {
	return bytesToHex(blake3(`${HANDLER}/c/${collectionId}/t/${tokenId}`))
}

export async function getNFT(bep131Store: KVStore, hash: string) {
	return BEP131Token.fromBinary(await bep131Store.get(hash), PROTO_OPTS)
}

export async function upsertNFT(bep131Store: KVStore, hash: string, nft: Plain<BEP131Token>) {
	const result = new BEP131Token(nft)

	await bep131Store.set(hash, result.toBinary())

	return result
}

export async function destroyNFT(bep131Store: KVStore, hash: string) {
	await bep131Store.forget(hash)
}
