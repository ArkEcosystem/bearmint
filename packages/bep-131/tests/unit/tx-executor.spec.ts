import { expect, describe, it, beforeEach } from 'vitest'
import { extendExpect } from '@bearmint/bep-005'
extendExpect(expect)
import type { TxTestContext } from '@bearmint/bep-006'
import { BEP131Token } from '@bearmint/bep-018'
import { getCollection, HANDLER } from '@bearmint/bep-126'

import { getNFT } from '../../source/utils.js'
import { createTx, setUp } from '../setup.js'

describe<TxTestContext>('Tx Executor', () => {
	setUp(beforeEach)

	it('should add the NFT to the account of the minter', async (context) => {
		const { sender, tx } = await createTx(context)

		expect(
			await context.state.getMultiStore().get('@bearmint/bep-126').getListHex(sender.address),
		).toStrictEqual(['a77d3cda10319ef1453ed69bfa8aef4c4c554a589d68a43d2e3eb63639bb3d85'])

		expect(
			await getCollection(
				context.state.getMultiStore().get('@bearmint/bep-126'),
				'a77d3cda10319ef1453ed69bfa8aef4c4c554a589d68a43d2e3eb63639bb3d85',
			),
		).toMatchSnapshot()

		await expect(await context.execute(tx)).toPassExecution()

		expect(
			await getCollection(
				context.state.getMultiStore().get('@bearmint/bep-126'),
				'a77d3cda10319ef1453ed69bfa8aef4c4c554a589d68a43d2e3eb63639bb3d85',
			),
		).toMatchSnapshot()

		expect(
			await context.state.getMultiStore().get('@bearmint/bep-126').getListHex(sender.address),
		).toStrictEqual(['a77d3cda10319ef1453ed69bfa8aef4c4c554a589d68a43d2e3eb63639bb3d85'])

		expect(
			await context.state.getMultiStore().get('@bearmint/bep-131').getListHex(sender.address),
		).toStrictEqual(['f013165441c2d9cc7ea26cfe5cd7d9cff50b06384a772d012ba92b0b65ca694a'])

		expect(
			await getNFT(
				context.state.getMultiStore().get('@bearmint/bep-131'),
				'f013165441c2d9cc7ea26cfe5cd7d9cff50b06384a772d012ba92b0b65ca694a',
			),
		).toStrictEqual(
			BEP131Token.fromJson({
				collection: 'a77d3cda10319ef1453ed69bfa8aef4c4c554a589d68a43d2e3eb63639bb3d85',
				id: 1,
				hash: 'f013165441c2d9cc7ea26cfe5cd7d9cff50b06384a772d012ba92b0b65ca694a',
				owner: sender.address,
			}),
		)
	})

	it('should add the NFT to the account of the recipient', async (context) => {
		const { recipient, sender, tx } = await createTx(context, undefined, true)

		expect(
			await context.state.getMultiStore().get('@bearmint/bep-126').getListHex(sender.address),
		).toStrictEqual(['a77d3cda10319ef1453ed69bfa8aef4c4c554a589d68a43d2e3eb63639bb3d85'])

		expect(
			await getCollection(
				context.state.getMultiStore().get('@bearmint/bep-126'),
				'a77d3cda10319ef1453ed69bfa8aef4c4c554a589d68a43d2e3eb63639bb3d85',
			),
		).toMatchSnapshot()

		await expect(await context.execute(tx)).toPassExecution()

		expect(
			await getCollection(
				context.state.getMultiStore().get('@bearmint/bep-126'),
				'a77d3cda10319ef1453ed69bfa8aef4c4c554a589d68a43d2e3eb63639bb3d85',
			),
		).toMatchSnapshot()

		expect(
			await context.state.getMultiStore().get('@bearmint/bep-126').getListHex(sender.address),
		).toStrictEqual(['a77d3cda10319ef1453ed69bfa8aef4c4c554a589d68a43d2e3eb63639bb3d85'])

		expect(
			await context.state.getMultiStore().get('@bearmint/bep-131').getListHex(recipient.address),
		).toStrictEqual(['f013165441c2d9cc7ea26cfe5cd7d9cff50b06384a772d012ba92b0b65ca694a'])

		expect(
			await getNFT(
				context.state.getMultiStore().get('@bearmint/bep-131'),
				'f013165441c2d9cc7ea26cfe5cd7d9cff50b06384a772d012ba92b0b65ca694a',
			),
		).toStrictEqual(
			BEP131Token.fromJson({
				collection: 'a77d3cda10319ef1453ed69bfa8aef4c4c554a589d68a43d2e3eb63639bb3d85',
				id: 1,
				hash: 'f013165441c2d9cc7ea26cfe5cd7d9cff50b06384a772d012ba92b0b65ca694a',
				owner: recipient.address,
			}),
		)
	})
})
