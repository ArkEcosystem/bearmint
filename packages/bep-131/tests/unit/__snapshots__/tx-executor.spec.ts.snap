// Vitest Snapshot v1

exports[`Tx Executor > should add the NFT to the account of the minter 1`] = `
BEP126Collection {
  "accessControlList": {
    "bear15z20cwcgszr54v7w7pxulyv3pyqjz878rmvzqtjf2sqphz7gv4hrz0f44vuwgx57qy6h0c064x3rw44zxpl": BEP126User {
      "permissions": [
        0,
        1,
        2,
      ],
      "role": 0,
    },
  },
  "currentOwner": "bear15z20cwcgszr54v7w7pxulyv3pyqjz878rmvzqtjf2sqphz7gv4hrz0f44vuwgx57qy6h0c064x3rw44zxpl",
  "id": "a77d3cda10319ef1453ed69bfa8aef4c4c554a589d68a43d2e3eb63639bb3d85",
  "initialOwner": "bear15z20cwcgszr54v7w7pxulyv3pyqjz878rmvzqtjf2sqphz7gv4hrz0f44vuwgx57qy6h0c064x3rw44zxpl",
  "name": "name",
  "policies": [
    BEP126Policy {
      "args": "",
      "name": "@bearmint/bep-145",
      "type": 0,
    },
    BEP126Policy {
      "args": "",
      "name": "@bearmint/bep-145",
      "type": 1,
    },
    BEP126Policy {
      "args": "",
      "name": "@bearmint/bep-145",
      "type": 2,
    },
  ],
  "symbol": "symbol",
  "tokens": [],
  "uriPrefix": "uriPrefix",
  "uriSuffix": "uriSuffix",
}
`;

exports[`Tx Executor > should add the NFT to the account of the minter 2`] = `
BEP126Collection {
  "accessControlList": {
    "bear15z20cwcgszr54v7w7pxulyv3pyqjz878rmvzqtjf2sqphz7gv4hrz0f44vuwgx57qy6h0c064x3rw44zxpl": BEP126User {
      "permissions": [
        0,
        1,
        2,
      ],
      "role": 0,
    },
  },
  "currentOwner": "bear15z20cwcgszr54v7w7pxulyv3pyqjz878rmvzqtjf2sqphz7gv4hrz0f44vuwgx57qy6h0c064x3rw44zxpl",
  "id": "a77d3cda10319ef1453ed69bfa8aef4c4c554a589d68a43d2e3eb63639bb3d85",
  "initialOwner": "bear15z20cwcgszr54v7w7pxulyv3pyqjz878rmvzqtjf2sqphz7gv4hrz0f44vuwgx57qy6h0c064x3rw44zxpl",
  "name": "name",
  "policies": [
    BEP126Policy {
      "args": "",
      "name": "@bearmint/bep-145",
      "type": 0,
    },
    BEP126Policy {
      "args": "",
      "name": "@bearmint/bep-145",
      "type": 1,
    },
    BEP126Policy {
      "args": "",
      "name": "@bearmint/bep-145",
      "type": 2,
    },
  ],
  "symbol": "symbol",
  "tokens": [
    "f013165441c2d9cc7ea26cfe5cd7d9cff50b06384a772d012ba92b0b65ca694a",
  ],
  "uriPrefix": "uriPrefix",
  "uriSuffix": "uriSuffix",
}
`;

exports[`Tx Executor > should add the NFT to the account of the recipient 1`] = `
BEP126Collection {
  "accessControlList": {
    "bear15z20cwcgszr54v7w7pxulyv3pyqjz878rmvzqtjf2sqphz7gv4hrz0f44vuwgx57qy6h0c064x3rw44zxpl": BEP126User {
      "permissions": [
        0,
        1,
        2,
      ],
      "role": 0,
    },
  },
  "currentOwner": "bear15z20cwcgszr54v7w7pxulyv3pyqjz878rmvzqtjf2sqphz7gv4hrz0f44vuwgx57qy6h0c064x3rw44zxpl",
  "id": "a77d3cda10319ef1453ed69bfa8aef4c4c554a589d68a43d2e3eb63639bb3d85",
  "initialOwner": "bear15z20cwcgszr54v7w7pxulyv3pyqjz878rmvzqtjf2sqphz7gv4hrz0f44vuwgx57qy6h0c064x3rw44zxpl",
  "name": "name",
  "policies": [
    BEP126Policy {
      "args": "",
      "name": "@bearmint/bep-145",
      "type": 0,
    },
    BEP126Policy {
      "args": "",
      "name": "@bearmint/bep-145",
      "type": 1,
    },
    BEP126Policy {
      "args": "",
      "name": "@bearmint/bep-145",
      "type": 2,
    },
  ],
  "symbol": "symbol",
  "tokens": [],
  "uriPrefix": "uriPrefix",
  "uriSuffix": "uriSuffix",
}
`;

exports[`Tx Executor > should add the NFT to the account of the recipient 2`] = `
BEP126Collection {
  "accessControlList": {
    "bear15z20cwcgszr54v7w7pxulyv3pyqjz878rmvzqtjf2sqphz7gv4hrz0f44vuwgx57qy6h0c064x3rw44zxpl": BEP126User {
      "permissions": [
        0,
        1,
        2,
      ],
      "role": 0,
    },
  },
  "currentOwner": "bear15z20cwcgszr54v7w7pxulyv3pyqjz878rmvzqtjf2sqphz7gv4hrz0f44vuwgx57qy6h0c064x3rw44zxpl",
  "id": "a77d3cda10319ef1453ed69bfa8aef4c4c554a589d68a43d2e3eb63639bb3d85",
  "initialOwner": "bear15z20cwcgszr54v7w7pxulyv3pyqjz878rmvzqtjf2sqphz7gv4hrz0f44vuwgx57qy6h0c064x3rw44zxpl",
  "name": "name",
  "policies": [
    BEP126Policy {
      "args": "",
      "name": "@bearmint/bep-145",
      "type": 0,
    },
    BEP126Policy {
      "args": "",
      "name": "@bearmint/bep-145",
      "type": 1,
    },
    BEP126Policy {
      "args": "",
      "name": "@bearmint/bep-145",
      "type": 2,
    },
  ],
  "symbol": "symbol",
  "tokens": [
    "f013165441c2d9cc7ea26cfe5cd7d9cff50b06384a772d012ba92b0b65ca694a",
  ],
  "uriPrefix": "uriPrefix",
  "uriSuffix": "uriSuffix",
}
`;
